includes/core/server/class-request.php

```diff
@@ -0,0 +1,523 @@
+	}
+
+	/**
+	 * Indicates if the raw response should be returned.
+	 *
+	 * @return bool
+	 */
+	public function should_return_raw_response(): bool {
+		return false;
+	}
+
+	/**
+	 * Returns all of the parameters for the request.
+	 *
+	 * @return array
+	 * @throws Invalid_Request_Parameter_Exception If the request has not been initialized yet.
+	 */
+	final public function get_params() {
+		$defaults = $this->get_default_params();
+		$params   = array_merge( $defaults, $this->params );
+
+		$missing_params = [];
+		foreach ( $this->get_required_params() as $name ) {
+			if ( ! isset( $params[ $name ] ) ) {
+				$missing_params[] = $name;
+			}
+		}
+
+		if ( ! empty( $missing_params ) ) {
+			throw new Invalid_Request_Parameter_Exception(
+				sprintf(
+					'Trying to access the parameters of a request which is not (fully) initialized yet. Missing parameter(s) for %s: %s',
+					get_class( $this ),
+					implode( ', ', $missing_params )
+				),
+				'wcpay_core_invalid_request_parameter_missing_parameters'
+			);
+		}
+
+		foreach ( $params as $key => $value ) {
+			if ( true === $value ) {
+				// The WCPay server requires the string 'true'.
+				$params[ $key ] = 'true';
+			}
+		}
+
+		return $params;
+	}
+
+	/**
+	 * Get request param by key.
+	 *
+	 * @param string $key Key to get.
+	 *
+	 * @return mixed
+	 * @throws Invalid_Request_Parameter_Exception
+	 */
+	final public function get_param( $key ) {
+		if ( array_key_exists( $key, $this->params ) ) {
+			return $this->params[ $key ];
+		}
+		throw new Invalid_Request_Parameter_Exception( 'The passed key does not exist in Request class', 'wcpay_core_invalid_request_parameter_uninitialized_param' );
+	}
+
+	/**
+	 * Allows the request to be modified, and then sends it.
+	 *
+	 * @param string $hook    The filter to use.
+	 * @param mixed  ...$args Other parameters for the hook.
+	 * @return mixed          Either the response array, or the correct object.
+	 */
+	final public function send( $hook, ...$args ) {
+		$request = $this->apply_filters( $hook, ...$args );
+
+		return $this->format_response(
+			$this->api_client->send_request( $request )
+		);
+	}
+
+	/**
+	 * Formats the response from the server.
+	 *
+	 * @param  mixed $response The response from `WC_Payments_API_Client::request`.
+	 * @return mixed           Either the same response, or the correct object.
+	 */
+	public function format_response( $response ) {
+		return new Response( $response );
+	}
+
+	/**
+	 * Stores a parameter within the internal props.
+	 *
+	 * Use this method within child classes in order to allow
+	 * those properties to be protected by overwriting.
+	 *
+	 * @param  string $key   The name of the parameter.
+	 * @param  mixed  $value And the value to set.
+	 * @return static        The instance of the class, ready for method chaining.
+	 */
+	final protected function set_param( string $key, $value ) {
+		if ( $this->protected_mode && in_array( $key, $this->get_immutable_params(), true ) ) {
```

Comment:
Could we split this check in two so we'd have different exception messages for each case?

Reply:
Maybe you are missing the point: The parameter is only immutable while the object is in protected mode.

Reply:
Yes, I was missing the point.